packages/pds/src/actor-store/repo/sql-repo-transactor.ts

```diff
@@ -0,0 +1,109 @@
+import { CommitData, RepoStorage, BlockMap } from '@atproto/repo'
+import { chunkArray } from '@atproto/common'
+import { CID } from 'multiformats/cid'
+import { ActorDb, RepoBlock } from '../db'
+import { SqlRepoReader } from './sql-repo-reader'
+
+export class SqlRepoTransactor extends SqlRepoReader implements RepoStorage {
+  cache: BlockMap = new BlockMap()
+  now: string
+
+  constructor(public db: ActorDb, now?: string) {
+    super(db)
+    this.now = now ?? new Date().toISOString()
+  }
+
+  // proactively cache all blocks from a particular commit (to prevent multiple roundtrips)
+  async cacheRev(rev: string): Promise<void> {
+    const res = await this.db.db
+      .selectFrom('repo_block')
+      .where('repoRev', '=', rev)
+      .select(['repo_block.cid', 'repo_block.content'])
+      .limit(15)
+      .execute()
+    for (const row of res) {
+      this.cache.set(CID.parse(row.cid), row.content)
+    }
+  }
+
+  async putBlock(cid: CID, block: Uint8Array, rev: string): Promise<void> {
+    this.db.assertTransaction()
+    await this.db.db
+      .insertInto('repo_block')
+      .values({
+        cid: cid.toString(),
+        repoRev: rev,
+        size: block.length,
+        content: block,
+      })
+      .onConflict((oc) => oc.doNothing())
+      .execute()
+    this.cache.set(cid, block)
+  }
+
+  async putMany(toPut: BlockMap, rev: string): Promise<void> {
+    this.db.assertTransaction()
+    const blocks: RepoBlock[] = []
+    toPut.forEach((bytes, cid) => {
+      blocks.push({
+        cid: cid.toString(),
+        repoRev: rev,
+        size: bytes.length,
+        content: bytes,
+      })
+      this.cache.addMap(toPut)
+    })
+    await Promise.all(
+      chunkArray(blocks, 500).map((batch) =>
+        this.db.db
+          .insertInto('repo_block')
+          .values(batch)
+          .onConflict((oc) => oc.doNothing())
+          .execute(),
+      ),
+    )
+  }
```

Comment:
It might be worth checking if 500 is actually possible here.  I think that sqlite maxes out at 999 bound parameters, and I could imagine there being more than that in this query.

Reply:
yeah good call. also i think sqlite might be happier with smaller chunk sizes anyway